core/trino-main/src/main/java/io/trino/execution/executor2/ThreadPerDriverTaskExecutor.java

```diff
@@ -0,0 +1,212 @@
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.DoubleSupplier;
+import java.util.function.Predicate;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class ThreadPerDriverTaskExecutor
+        implements TaskExecutor
+{
+    private final FairScheduler scheduler;
+    private final Tracer tracer;
+    private final VersionEmbedder versionEmbedder;
+    private volatile boolean closed;
+
+    @Inject
+    public ThreadPerDriverTaskExecutor(TaskManagerConfig config, Tracer tracer, VersionEmbedder versionEmbedder)
+    {
+        this(tracer, versionEmbedder, new FairScheduler(config.getMaxWorkerThreads(), "SplitRunner-%d", Ticker.systemTicker()));
+    }
+
+    @VisibleForTesting
+    public ThreadPerDriverTaskExecutor(Tracer tracer, VersionEmbedder versionEmbedder, FairScheduler scheduler)
+    {
+        this.scheduler = scheduler;
+        this.tracer = requireNonNull(tracer, "tracer is null");
+        this.versionEmbedder = requireNonNull(versionEmbedder, "versionEmbedder is null");
+    }
+
+    @PostConstruct
+    @Override
+    public synchronized void start()
+    {
+        scheduler.start();
+    }
+
+    @PreDestroy
+    @Override
+    public synchronized void stop()
+    {
+        closed = true;
+        scheduler.close();
+    }
+
+    @Override
+    public synchronized TaskHandle addTask(
+            TaskId taskId,
+            DoubleSupplier utilizationSupplier,
+            int initialSplitConcurrency,
+            Duration splitConcurrencyAdjustFrequency,
+            OptionalInt maxDriversPerTask)
+    {
+        checkArgument(!closed, "Executor is already closed");
+
+        Group group = scheduler.createGroup(taskId.toString());
+        return new TaskEntry(taskId, group);
+    }
+
+    @Override
+    public synchronized void removeTask(TaskHandle handle)
+    {
+        TaskEntry entry = (TaskEntry) handle;
+
+        if (!entry.isDestroyed()) {
+            scheduler.removeGroup(entry.group());
+            entry.destroy();
+        }
+    }
+
+    @Override
+    public synchronized List<ListenableFuture<Void>> enqueueSplits(TaskHandle handle, boolean intermediate, List<? extends SplitRunner> splits)
+    {
+        checkArgument(!closed, "Executor is already closed");
+
+        TaskEntry entry = (TaskEntry) handle;
+
+        List<ListenableFuture<Void>> futures = new ArrayList<>();
+        for (SplitRunner split : splits) {
+            entry.addSplit(split);
+
+            int splitId = entry.nextSplitId();
+            ListenableFuture<Void> done = scheduler.submit(entry.group(), splitId, new VersionEmbedderBridge(versionEmbedder, new SplitProcessor(entry.taskId(), splitId, split, tracer)));
+            done.addListener(split::close, directExecutor());
+            futures.add(done);
```

Comment:
No, it can't throw unless the scheduler is shut down (which can only happen on `stop()` and it would be caught by the check at the top of this method), or if something is very broken.